x-pack/plugins/ml/public/application/app.tsx

```diff
@@ -53,13 +56,20 @@ const localStorage = new Storage(window.localStorage);
 /**
  * Provides global services available across the entire ML app.
  */
-export function getMlGlobalServices(httpStart: HttpStart, usageCollection?: UsageCollectionSetup) {
+export function getMlGlobalServices(
+  httpStart: HttpStart,
+  dataViews: DataViewsContract,
+  usageCollection?: UsageCollectionSetup
+) {
   const httpService = new HttpService(httpStart);
   const mlApiServices = mlApiServicesProvider(httpService);
+  const mlIndexUtils = indexServiceFactory(dataViews);
+  const mlFieldFormatService = fieldFormatServiceFactory(mlApiServices, mlIndexUtils);
```

Comment:
To prevent bloat of the context, perhaps the `mlFieldFormatService` should be created in the component as needed and then passed/made available to it's children? 

For example, we already have `mlApiServices`  and `dataViews` in context, I believe.
 `fieldFormatServiceFactory` requires `mlApiServices` and `mlIndexUtils` - which, in turn,  only requires the dataViewsService.
 
 From that, fieldFormatService could be instantiated once in the component in which it is needed instead of adding more to the context.

Reply:
Not sure it's bloat? Prior to this PR the service was only instantiated once too, it's just that with the dependency cache under the hood it could be imported everywhere without being passed around via context. So this variant felt closer to the original then instantiating multiple times. Note the factory doesn't have any costly side effects, for example it doesn't call any requests, that's only happening once you call one of its methods.

Reply:
I tend to agree that it would be better to keep `mlIndexUtils` and `mlFieldFormatService` out of this context as they are used so rarely throughout our plugin. It sets a precedence that all utils and services we have should go in here too.
Instantiating them when needed in each file would be better IMO.

Reply:
As discussed offline, I'm leaving `mlFieldFormatService` here as a global service for now because of its internal state. `mlIndexUtils` will not be passed on to global services, instead it can be accessed within components via `useMlIndexUtils()`.

I added the following comment to each place where the services are set up (the main ML app, the anomaly charts embeddable and the single metric viewer embeddable):

```
    // Note on the following services:
    // - `mlIndexUtils` is just instantiated here to be passed on to `mlFieldFormatService`,
    //   but it's not being made available as part of global services. Since it's just
    //   some stateless utils `useMlIndexUtils()` should be used from within components.
    // - `mlFieldFormatService` is a stateful legacy service that relied on "dependency cache",
    //   so because of its own state it needs to be made available as a global service.
    //   In the long run we should again try to get rid of it here and make it available via
    //   its own context or possibly without having a singleton like state at all, since the
    //   way this manages its own state right now doesn't consider React component lifecycles.
```

Reply:
Does this still require a code change? `mlIndexUtils` is still being added to the context.

Reply:
Oh yes, overlooked that one, thanks! Fixed in bccd244.